src/traces/sankey/calc.js

```diff
@@ -34,7 +36,32 @@ function convertToD3Sankey(trace) {
         components[cscale.label] = scale;
     }
 
-    var nodeCount = nodeSpec.label.length;
+    var maxNodeId = 0;
+    for(i = 0; i < linkSpec.value.length; i++) {
+        if(linkSpec.source[i] > maxNodeId) maxNodeId = linkSpec.source[i];
+        if(linkSpec.target[i] > maxNodeId) maxNodeId = linkSpec.target[i];
+    }
+    var nodeCount = maxNodeId + 1;
+
+    // Group nodes
+    var j;
+    var groups = trace.node.groups;
+    var groupLookup = {};
+    for(i = 0; i < groups.length; i++) {
+        var group = groups[i];
+        // Build a lookup table to quickly find in which group a node is
+        if(Array.isArray(group)) {
+            for(j = 0; j < group.length; j++) {
+                var nodeIndex = group[j];
+                var groupIndex = nodeCount + i;
+                groupLookup[nodeIndex] = groupIndex;
+            }
+        } else {
+            Lib.warn('node.groups must be an array, default to empty array []');
```

Comment:
If we declare `groups` an `info_array`, we wouldn't need this. `Plotly.validate` would be enough to catch bad inputs.

Reply:
Nice!  :hocho: in 1cd206f7fdba75b010650331fc1f7a726ec488df